torch/_subclasses/fake_tensor.py

```diff
@@ -51,6 +52,90 @@ def _is_tensor_constructor(func: OpOverload):
 # FakeTensor extends MetaTensors to also carry an additional `fake_device`
 # which tracks devices that would have been used.
 
+def torch_dispatch_impl(cls_or_mode_instance, func, types, args, kwargs, run_function):
+    kwargs = kwargs if kwargs else {}
+
+    # This classes virtualizes .device() calls, need to short-circuit
+    # it insteead of calling device again or we would keep on recurring
+    if func == torch.ops.prim.device.default:
+        assert len(args) == 1 and isinstance(args[0], FakeTensor)
+        return args[0].fake_device
+
+    def wrap(e, device=None):
+        if isinstance(e, torch.Tensor) and not isinstance(e, FakeTensor):
+            if device:
+                return FakeTensor(e, device)
```

Comment:
should be in `no_dispatch` to allow for nested modes (more info https://github.com/pytorch/pytorch/pull/78516/files#r885905994)